aggregator/src/diagnostic.rs

```diff
@@ -0,0 +1,123 @@
+//! Writing diagnostic files to disk.
+
+use anyhow::Context;
+use derivative::Derivative;
+use janus_messages::{AggregationJobId, ReportMetadata};
+use std::{fmt::Debug, fs::File, io::Write, path::Path, time::SystemTime};
+use uuid::Uuid;
+
+/// Represents an illegal attempt to mutate an aggregation job.
+#[derive(Derivative, Clone)]
+#[derivative(Debug)]
+pub struct AggregationJobInitForbiddenMutationEvent {
+    /// The ID of the aggregation job.
+    pub aggregation_job_id: AggregationJobId,
+
+    /// The SHA-256 of the request that created the aggregation job.
+    #[derivative(Debug(format_with = "fmt_hash_option"))]
+    pub original_request_hash: Option<[u8; 32]>,
+
+    /// The ordered report metadatas from the request that created the aggregation job.
+    pub original_report_metadatas: Vec<ReportMetadata>,
+
+    /// The batch ID in the original request.
+    pub original_batch_id: String,
+
+    /// The aggregation param in the original request.
+    pub original_aggregation_parameter: Vec<u8>,
+
+    /// The SHA-256 of the request that attempted to mutate the aggregation job.
+    #[derivative(Debug(format_with = "fmt_hash_option"))]
+    pub mutating_request_hash: Option<[u8; 32]>,
+
+    /// The ordered report metadatas from the request that attempted to mutate the aggregation job.
+    pub mutating_request_report_metadatas: Vec<ReportMetadata>,
+
+    /// The batch ID of the mutating request.
+    pub mutating_request_batch_id: String,
+
+    /// The aggregation param in the mutating request.
+    pub mutating_request_aggregation_parameter: Vec<u8>,
+}
+
+fn fmt_hash_option(
+    v: &Option<[u8; 32]>,
+    f: &mut std::fmt::Formatter,
+) -> Result<(), std::fmt::Error> {
+    write!(f, "{:?}", v.map(hex::encode))
+}
+
+/// Write an event.
+///
+/// The [`std::fmt::Debug`] representation of `event` will be written to a file named
+/// `<event_type>-<timestamp>-<ID>-<version>`, where `event_type` is the provided argument,
+/// `timestamp` is the time at which the event was recorded, `ID` is a UUIDv4 and `version` is the
+/// version of Janus that generated the event.
+pub fn write_event<P: AsRef<Path>>(
+    event_storage: P,
+    event_type: &'static str,
+    event: Box<dyn Debug>,
+) -> Result<Uuid, anyhow::Error> {
+    let now = SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .context("failed to get current time")?;
+    let event_id = Uuid::new_v4();
+    let version = env!("CARGO_PKG_VERSION");
+
+    let mut f = File::create(
+        event_storage
+            .as_ref()
+            .join(format!("{event_type}-{now:?}-{event_id}-{version}")),
+    )
+    .context("failed to open file to write event")?;
+
+    write!(&mut f, "{event:#?}").context("failed to write event")?;
```

Comment:
It might be good to call `sync_all()` to catch possible errors when syncing and closing as well.

Separately, since we expect to be writing at least 10MiB of formatted text, this call could take a while. I don't know what sort of latency we would expect from file creation in the FUSE driver, for example. Thus, I think we should make `write_event()` an async method, and then use `spawn_blocking()` around creating the file, writing to it, and closing it. (The only good stories for heavy-duty filesystem work in Tokio are spawn_blocking with standard syscalls or an experimental io_uring crate)

Reply:
Done. I considered using `tokio::fs::File` for this, but it's not obvious to me how to use an async writer with `serde_json`, so I stuck with the synchronous `std::fs::File`.